### PR TITLE
Add face-tracking support

### PR DESCRIPTION
**Pull request by cjbarth**
# Pull request

## Type of change

This is a PR to address the feature request in #316 

## Summary

This builds upon the work in FaceDetect to track faces by zooming and panning based on the detected faces.

## Related Issue

#316 

## More info

There are several configuration options to change the face detection method, the zoom rate, the aspect ratio, and the margin/padding applied to the detected face.

## Added dependencies

No dependencies added.

## Target Environment

This shouldn't change any target environment requirements.

### CODE REVIEW

**Comment by hipersayanX** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp  6
- libAvKys/Plugins/FaceDetect/src/facedetectelement.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/webcamoid/webcamoid/pullRequest?prid=5768805&bid=18879631)

**Comment by hipersayanX** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp  6
- libAvKys/Plugins/FaceDetect/src/facedetectelement.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/webcamoid/webcamoid/pullRequest?prid=5768805&bid=18879631)

**Comment by hipersayanX** [comment]
I like it, pretty cool!
The code is ok in general, there are just a few things to fix.

# The code

For example in [facedetectelement](https://github.com/cjbarth/webcamoid/blob/face-tracking/libAvKys/Plugins/FaceDetect/src/facedetectelement.h), instead of creating a **face** and **detectOnly** property and a calling iStream and so on, just create a QVector<QRect> detectFaces(AkVideoPacket), so you invoke detectFaces, get the result, and call setFaces with the result straightforward, instead of, [calling iStream](https://github.com/cjbarth/webcamoid/blob/94c2a4b3d408cdd066504563d0c3b2c248ada39a/libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp#L233), then [receiving the results in a slot](https://github.com/cjbarth/webcamoid/blob/94c2a4b3d408cdd066504563d0c3b2c248ada39a/libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp#L336), process the results and going back [continuing the program after iStream](https://github.com/cjbarth/webcamoid/blob/94c2a4b3d408cdd066504563d0c3b2c248ada39a/libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp#L234) (because all of that code is processed in the same thread), that makes a bit hard to follow the code.

Coding style also needs a some cleaning.
* [if, for, while, switch, etc., needs 1 new line before and after](https://github.com/cjbarth/webcamoid/blob/face-tracking/libAvKys/Plugins/FaceDetect/src/facedetectelement.cpp#L206)
* [This function is private and can't be used outside, move to FaceTrackElementPrivate instead](https://github.com/cjbarth/webcamoid/blob/94c2a4b3d408cdd066504563d0c3b2c248ada39a/libAvKys/Plugins/FaceTrack/src/facetrackelement.h#L100)
* [Unnecessary new line](https://github.com/cjbarth/webcamoid/blob/94c2a4b3d408cdd066504563d0c3b2c248ada39a/libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp#L47)
* [Lines 56 to 59 can be moved to FaceTrackElementPrivate](https://github.com/cjbarth/webcamoid/blob/face-tracking/libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp#L56)
* [Respect 80 char lines limit, break one long statement into many small statements if necessary](https://github.com/cjbarth/webcamoid/blob/face-tracking/libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp#L138)
* [Also, don't forget to add your copyright](https://github.com/cjbarth/webcamoid/blob/94c2a4b3d408cdd066504563d0c3b2c248ada39a/libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp#L2)

You can check all coding style rules [here](https://github.com/webcamoid/webcamoid/wiki/Coding-style-and-conventions).

# Plugin controls

* It took me a bit to understand the plugin because "it did not worked", or at least on how it was configured by default. I would change all padding and margins values to maybe 10, because the default values did not worked well.
* Change face bracketing duration to 1 so the changes shows faster.
* You have hard-coded the if-else that draw the buckets, instead add an option to show/hide the buckets, I think the users and specially other developers would be interested on seeing the algorithm in action.
* Also, would be possible to take the aspect ratio from the incoming frame instead of making it user defined?

But again, in general I loved it, I was playing with it for a long time :smile:

**Comment by cjbarth** [comment]
Thank you so much for your thorough review and kind manner. I will take your suggestions and incorporate them as best I can. I just a have a few comments in reply.

> I loved it

YAY!!! I can't get the installer to compile on my machine, so I can't actually test this with any video conferencing software, so I'm anxious to finish this and get it in to an official build so I can use it. Honestly, I also realized there is probably a need for a very simple "crop" plugin, so I might work on that next.

> just create a QVector detectFaces(AkVideoPacket), so you invoke detectFaces, get the result

Honestly, I'd love to, but I can't figure out how. I can't seem to call anything in another plugin that isn't part of `AkElement`. I can get or set a property, but I can't call a function with arguments. I tried several different ways. This must be my lack of experience with Qt, C++ (I haven't done C++ in a long while), or this specific project. Even checking the other times that plugs-ins are called from each other, I see no instances of when a function was called that isn't part of the interface declared by `AkElement`.

>     * You have hard-coded the if-else that draw the buckets

I was intending to just remove that code completely when I got everything working, and therefore I didn't spend much time making the boxes make much sense. If this is a feature that you'd actually like to see, what would you suggest for making the concept of the boxes clearer? They way it works now is to rotate the color and line thickness to help identify which bucket is represented by the box, but I'm not sure that would be helpful to anyone not stepping through the code.

>     * Change face bracketing duration to 1 so the changes shows faster.

I tested this way myself, but found that during a normal 'session' sitting in front of the camera, the face detection was a little too jumpy and the panning and zooming would happen a little too often. How might I recommend a good setting for usage vs. a setting for seeing the effects immediately?

>     * I would change all padding and margins values to maybe 10, because the default values did not worked well.

The defaults were configured to work with all sorts of aspect ratios and to make sure that face, forehead, ears, and shoulders were all visible in the frame, for all faces detected in the frame. What were you seeing that indicated to you that it wasn't working? Since I only have a test case of one webcam, I'm can see how there might be room for improvement, however, for a presentation, the zoom should never be just to the face.

>     * would be possible to take the aspect ratio from the incoming frame instead of making it user defined

Probably, but it seems to me that most conferencing software expects a 16:9 output and the way that Zoom, Skype, Teams, etc. are set up all show multiple participants best with the source is 16:9. Should I include a button next to the box to 'reset' to the camera default?

**Comment by hipersayanX** [comment]
>I can't get the installer to compile on my machine

Which operating system? What problems did you had?

>Honestly, I'd love to, but I can't figure out how. 

You have an example [here](https://github.com/webcamoid/webcamoid/blob/b8efd2b2636f5b3396c2756cdfa6177ec9320911/libAvKys/Plugins/VideoCapture/src/videocaptureelement.h#L69) and [here](https://github.com/webcamoid/webcamoid/blob/b8efd2b2636f5b3396c2756cdfa6177ec9320911/StandAlone/src/videolayer.cpp#L682), the method must be public otherwise you won't be able to call it outside the plugin.

>I was intending to just remove that code completely when I got everything working

Nah, leave it, I like it.

>what would you suggest for making the concept of the boxes clearer? They way it works now is to rotate the color and line thickness to help identify which bucket is represented by the box

I was thinking on assigning the colors based on a [hue rotation](https://es.wikipedia.org/wiki/Archivo:RGB_color_circle.png), let say you have a bucket of 4, you divide the circle in 4 slices, then you assign red to the first rectangle (the inner one), green to the second, cyan to to third, and purple to fourth, and so on. And maybe also assigning an opacity: 100%, 75%, 50% and 25%.
The thickness could be configurable by the user and default to 2 maybe.

>I tested this way myself, but found that during a normal 'session' sitting in front of the camera, the face detection was a little too jumpy and the panning and zooming would happen a little too often. How might I recommend a good setting for usage vs. a setting for seeing the effects immediately?

The problems is more like the first time usage impact, and the user thinking "was does this thing do?", an experienced user will know what he need to modify in order to get the plugin working as he want, but someone who doesn't know how it works won't be able to understand it. Also due to this, it may be unlocked by default.

>What were you seeing that indicated to you that it wasn't working?

A was seeing the frame as is, or as much going up, I was expecting the frame getting closer to my face or the plugin following me if I moved left or right. But again, is more like a question of expectation.

>Should I include a button next to the box to 'reset' to the camera default?

Maybe a switch, when it is disabled it will take the aspect ratio from the frame and disable the manual one, when enabled it will take the manual one.

Also I was thinking, you have two variables: padding and margin. Padding is the extra space between the detected rectangle and it's internal contents, margin is the extra space between the detected rectangle and the outside content, but at the end what you will draw is the area covered by the margin rectangle, so you can just leave the margin variable. For example, if you have a margin of 10% and a padding of 5%, is the same as adding a margin of 15% to the detected frame (the detected frame is the [content](https://commons.wikimedia.org/wiki/File:W3C_and_Internet_Explorer_box_models.svg)). Same effect with simplified controls.

**Comment by cjbarth** [comment]
> You have an example [here](https://github.com/webcamoid/webcamoid/blob/b8efd2b2636f5b3396c2756cdfa6177ec9320911/libAvKys/Plugins/VideoCapture/src/videocaptureelement.h#L69) and [here](https://github.com/webcamoid/webcamoid/blob/b8efd2b2636f5b3396c2756cdfa6177ec9320911/StandAlone/src/videolayer.cpp#L682), the method must be public otherwise you won't be able to call it outside the plugin.

I tried that and got `QMetaMethod::invoke: Unable to invoke methods with return values in queued connections`. So, the best I can do is to call the method and then let the results be processed in a slot. So, I figured, why go through all that and end up with all that duplicated code, so I used the method I did. What might I be doing wrong?

**Comment by hipersayanX** [comment]
Call it with [Qt::DirectConnection](https://doc.qt.io/qt-5/qmetaobject.html#invokeMethod).

**Comment by hipersayanX** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- libAvKys/Plugins/FaceDetect/src/facedetectelement.cpp  1
- libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/webcamoid/webcamoid/pullRequest?prid=5768805&bid=18879631)

**Comment by cjbarth** [comment]
I've pushed up another draft taking in to consideration many of the suggestions I've received. I think there might still be some code-style issues to clean up, but I was wondering, is there a linter for that?

I've tried to improve the sensibility of the defaults to give a good user experience, balancing between the obviousness of this working as intended and not being jerky to the point of being unusable due to the fuzziness of the face detection algorithms.

Regarding the merging of padding and margin, that really isn't possible as it would break the function and yield different behavior. Padding is added to the detected face before it is saved. Margin is added to the face temporarily when deciding if the newly detected face is different enough to be worth saving; margin is not saved and not rendered with the debug boxes. It did occur to me though that margin might be a value that can be simplified down to a single number and just have that apply to all 4 sides, but, I'm not sure if the simplification of the UI is worth the reduction in configurability.

Please have another look at this an offer additional feedback. Once the functionality is good, I'll mark the PR as ready for review.

**Comment by hipersayanX** [comment]
>I've pushed up another draft taking in to consideration many of the suggestions I've received. I think there might still be some code-style issues to clean up, but I was wondering, is there a linter for that?

[This](https://gist.github.com/hipersayanX/25d9fa77cc622b86cbfc537c4162fda2) is the settings that I'm using (_Menu > Tools > Options > C++ > Code Style_), but it does not covers all cases, it does not covers the 80 lines limit and block separation, but anyway you can check any code example inside the project, [this](https://github.com/webcamoid/webcamoid/blob/master/libAvKys/Plugins/Denoise/src/denoiseelement.cpp#L259) for example.

>I've tried to improve the sensibility of the defaults to give a good user experience, balancing between the obviousness of this working as intended and not being jerky to the point of being unusable due to the fuzziness of the face detection algorithms.

Much better.

>Regarding the merging of padding and margin, that really isn't possible as it would break the function and yield different behavior.

Ok, no problem, that's fine.

>Please have another look at this an offer additional feedback. Once the functionality is good, I'll mark the PR as ready for review.

Just align the [checkboxes](https://github.com/cjbarth/webcamoid/blob/face-tracking/libAvKys/Plugins/FaceTrack/share/qml/main.qml#L394) to the right (fo UI consistency) and clean the code, and I'll merge it.

**Comment by cjbarth** [comment]
I did my best with the white-space without sacrificing readability (too much). Please let me know if there are other formatting changes you'd like to see.

**Comment by hipersayanX** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- libAvKys/Plugins/FaceTrack/src/facetrackelement.cpp  2
- libAvKys/Plugins/FaceDetect/src/facedetectelement.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/webcamoid/webcamoid/pullRequest?prid=5768805&bid=18879631)

**Comment by hipersayanX** [comment]
I merged it, but I had to change many things, check 77eaeac.

**Comment by cjbarth** [comment]
Thank you for merging it. I'm sorry about all the changes you had to make. I would have made the changes, but I understand how it was probably quicker for you just to make the changes yourself. I hope it works well for everyone.
